microprofile/WFLY-12313_MicroProfile_OpenAPI_1.1.adoc

```diff
@@ -0,0 +1,126 @@
+= Integrate MicroProfile OpenAPI 1.1
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This enhancement proposes to add support for https://download.eclipse.org/microprofile/microprofile-open-api-1.1.2/microprofile-openapi-spec.html[Eclipse MicroProfile OpenAPI 1.1] to WildFly by integrating the implementation from https://github.com/smallrye/smallrye-open-api[SmallRye].
+This spec provides a unified Java API for producing documentation that complies with the https://github.com/OAI/OpenAPI-Specification/blob/3.0.2/versions/3.0.2.md[OpenAPI v3 specification].
+
+An MicroProfile OpenAPI subsystem exposes an HTTP GET `/openapi` endpoint that returns an OpenAPI v3 compliant document (in YAML or JSON format) containing documentation of the RESTful services of an application.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12313
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+The SmallRye implementation of the MicroProfile OpenAPI specification is already integrated by the following projects:
+
+* https://thorntail.io/[Thorntail] via the https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-openapi[microprofile-openapi fraction].
+* https://quarkus.io/[Quarkus] via the https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-openapi[smallrye-openapi extension].
+
+These integrations will heavily inform, though not dictate, the corresponding integration into WildFly.
+
```

Comment:
@pferraro, given what in the above resolved discussions, can we consolidate the snippet about migrating sample Thorntail app to MP featured Wildfly by just showing what @rsvoboda suggested (dependency modification in POM)? This would add great value to the users experience by showing them how concretely Wildfly would replace Thorntail.

Reply:
@fabiobrz What "dependency modification in POM" are you talking about?

Reply:
Hi @pferraro, I was referring to the task of migrating from Thorntail topic. 
I've just created a basic JAX-RS Thorntail example [1] following the guide and then "migrated" it to Wildfly [2]. 
Of course, these are just samples and the goal is just to show that a [very simple] Thorntail app can be migrated easily to Wildfly by making some changes to the POM. I've been testing [2] against current feature branch, [3].

[1]
https://github.com/fabiobrz/openapi-demo

[2]
https://github.com/fabiobrz/openapi-demo-wfly

[3]
https://github.com/MikeEdgar/wildfly/tree/feature/WFLY-12313-openapi

Reply:
@fabiobrz I'm still not sure what migration there is to do.  The only dependency such an app would have has not changed.

Reply:
What would the "only dependency" be? 
I see I had to remove all of Thorntail deps as the subsystem would provide module. 
Isn't that the migration change on the POM we're discussing about?

Reply:
That has to do with Thorntail provisioning, not the application itself.  Your application itself would have no references to any TT modules.  An application destined to run on WildFly has a completely different workflow.

Reply:
>  That has to do with Thorntail provisioning, not the application itself.

I don't think so. Probably - with _provisioning_ - you're referring to the plugin, which would execute the app by running the following command: 
```
mvn thorntail:run
```

But I wouldn't assume that's every application case. Quoting the project page:
```
Thorntail offers an innovative approach to _packaging_ and running Java EE applications by packaging them with just enough of the server runtime to "java -jar" your application. It's MicroProfile compatible, too...
```
So isn't TT one of the [currently] available tools that I actually need to have the feature? 
I think so. Otherwise I'd have to choose OpenLiberty or do something else. 
In such case, then, isn't adding the TT dependencies what would activate the needed fraction(s)? I think so, it is what I need to let a JVM execute my `uber jar` by running the commands:
```
mvn package
java -jar myapp-thorntail.war
```

> Your application itself would have no references to any TT modules.

My TT app would, indeed. My Wildfly app wouldn't, instead. And that's the whole point of migrating.
I am referring to what on TT project page, "Simple Project Setup" section: https://thorntail.io/
From that section I understand that my TT application itself should indeed have at least to depend on the TT bom (see TT examples) and also wouldn't have any MicroProfile OpenAPI feature if I wouldn't have the fraction declared, right? Quoting the snippet description there:
```
Import the BOM (Bill of Materials) dependency
...
Declare the fractions you need (or not, we’ll do our best to auto-detect them for you):
...
```
(I guess I would declare them in to be on the safe side a real world production scenario).

> An application destined to run on WildFly has a completely different workflow.

Indeed, that is quite clear but doesn't change the matter: a TT application would need dependencies to activate the fraction that support the feature, while an application that would run on Wildfly will have the feature thanks to a subsystem backed by a module and wouldn't need any MP dependency.

Please correct in case I am wrong on what above as I am no TT expert, as you know.
But if what I reported makes sense to you, too, then my point stands: 
"Do you have a TT app with MP fraction to support OpenAPI generation and you want to migrate that very app to Wildfly? Then remove the TT dependencies and let the Wildfly subsystem provide the same feature to you".

Reply:
I thinking I'm just tripping over terminology.  IMO, the uber jar isn't the application.  It's a jar containing both the application and the provisioned TT fractions.
General documentation about migrating TT provisioning to WF is a completely separate topic. I don't think it makes sense to have to repeat these instructions for every MP module.  However, to your point, general migration from TT is certainly something that needs documentation.   We should file a separate JIRA issue to track this migration documentation task.

Reply:
> General documentation about migrating TT provisioning to WF is a completely separate topic. I don't think it makes sense to have to repeat these instructions for every MP module. However, to your point, general migration from TT is certainly something that needs documentation. We should file a separate JIRA issue to track this migration documentation task

@pferraro that sounds reasonable to me as well. I think we could place a link for the JIRA once it gets created, so that it is at least traced here, WDYT?